distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso

```diff
@@ -383,57 +388,33 @@ Text.split self delimiter="," matcher=Text_Matcher.Case_Sensitive = if delimiter
-         "aaa aaa".replace "aa" "c" mode=Matching_Mode.Last matcher=Regex_Matcher . should_equal "aaa ca"
-Text.replace : Text -> Text -> Matching_Mode | Regex_Mode -> (Text_Matcher | Regex_Matcher) -> Text
-Text.replace self term="" new_text="" mode=Regex_Mode.All matcher=Text_Matcher.Case_Sensitive = if term.is_empty then self else
-    case matcher of
-        _ : Text_Matcher ->
+     Regexp replace.
+
+     '<a href="url">content</a>'.replace '<a href="(.*?)">(.*?)</a>' '$2 is at $1' use_regex=True == 'content is at url'
+
+Text.replace : Text -> Text -> Boolean -> Boolean -> Text
+Text.replace self term replacement case_sensitivity=Case_Sensitivity.Sensitive only_first=False use_regex=False = if term.is_empty then self else
+    case use_regex of
+        False ->
             array_from_single_result result = case result of
                 Nothing -> Array.empty
                 _ -> Array.new_1 result
-            spans_array = case matcher of
-                Text_Matcher.Case_Sensitive -> case mode of
-                    Regex_Mode.All ->
-                        Text_Utils.span_of_all self term
-                    Matching_Mode.First ->
-                        array_from_single_result <| Text_Utils.span_of self term
-                    Matching_Mode.Last ->
-                        array_from_single_result <| Text_Utils.last_span_of self term
-                    _ -> Error.throw (Illegal_Argument.Error "Invalid mode.")
-                Text_Matcher.Case_Insensitive locale -> case mode of
-                    Regex_Mode.All ->
+            spans_array = case case_sensitivity of
+                Case_Sensitivity.Sensitive -> case only_first of
+                    False -> Text_Utils.span_of_all self term
+                    True -> array_from_single_result <| Text_Utils.span_of self term
+                Case_Sensitivity.Insensitive locale -> case only_first of
+                    False ->
                         Text_Utils.span_of_all_case_insensitive self term locale.java_locale
-                    Matching_Mode.First ->
+                    True ->
                         array_from_single_result <|
                             Text_Utils.span_of_case_insensitive self term locale.java_locale False
-                    Matching_Mode.Last ->
-                        array_from_single_result <|
-                            Text_Utils.span_of_case_insensitive self term locale.java_locale True
-                    _ -> Error.throw (Illegal_Argument.Error "Invalid mode.")
-            Text_Utils.replace_spans self spans_array new_text
-        _ : Regex_Matcher ->
-            compiled_pattern = matcher.compile term
-            compiled_pattern.replace self new_text mode=mode
+            Text_Utils.replace_spans self spans_array replacement
+        True ->
+            case_insensitive = case_sensitivity.is_case_insensitive_in_memory
+            pattern = Regex_2.compile term case_insensitive=case_insensitive
```

Comment:
`Case_Sensitivity.Insensitive` has the `locale` parameter, but here we are ignoring it.

This may lead to unexpected behaviour if someone overrides the locale but that does not change the Regex behaviour.

It is ok if our Regex implementation cannot support custom locales. But IMO we should be explicit about it.

Let's add something similar to `assume_default_locale` (can be found in `Database`) here so that this will work as-is if the default locale was provided (99.999% of the time probably) but if the locale was changed - let's be explicit that only the default is supported and throw `Illegal_Argument.Error`.

Reply:
Regex does not support locale - it is strictly Unicode folding.

We should indeed throw an Illegal_Argument error here if not default.